src/java/main/org/apache/zookeeper/server/quorum/QuorumCnxManager.java

```diff
@@ -647,11 +648,11 @@ public void run() {
                         numRetries = 0;
                     }
                 } catch (IOException e) {
-                    if (shutdown) {
-                        break;
-                    }
                     LOG.error("Exception while listening", e);
-                    numRetries++;
+                    if (!(e instanceof SocketTimeoutException)) {
+                        numRetries++;
+                    }
+                }finally {
```

Comment:
it's my first time to commit code on github. i open a new  [pr](https://github.com/apache/zookeeper/pull/336), which has fixed according to your opinion.  I am sorry for my inconvenience.